File: Prasad_Bhagwat_ModelBasedCF.scala
// Imports required for the program
import org.apache.spark.{SparkConf, SparkContext}
import java.io._
import org.apache.spark.mllib.recommendation.{ALS, Rating}


object ModelBasedCF{

  // Counting error buckets
  def count_item_differences(actual: Float, predicted: Float): String= {
    val error = (math.abs(actual - predicted)).toFloat

    if (error >= 0.toFloat && error < 1.toFloat) {
      ">= 0 and <1"
    }
    else if (error >= 1.toFloat && error < 2.toFloat) {
      ">=1 and <2"
    }
    else if (error >= 2.toFloat && error < 3.toFloat) {
      ">=2 and <3"
    }
    else if (error >= 3.toFloat && error < 4.toFloat) {
      ">=3 and <4"
    }
    else {
      ">=4"
    }
  }

  // Main Function
  def main(args: Array[String]){
    val start_time = System.nanoTime()

    // Command Line Arguments
    val input_file      = args(0)
    val testing_file    = args(1)

    // Output filename generation
    val output_file_name    = "Prasad_Bhagwat_ModelBasedCF.txt"
    val output_file         = new PrintWriter(new File(output_file_name))

    // Creating Spark Context
    val spark_config    = new SparkConf()
    val spark_context   = new SparkContext(spark_config)
    spark_context.setLogLevel("WARN")

    // Reading input training data file and extracting header
    val input           = spark_context.textFile(input_file).filter(x => ! x.contains("userId"))
    val input_data      = input.map( x => {
                            val y = x.split(",")
                            ((y(0).toInt, y(1).toInt), y(2).toFloat)
                          })

    // Reading test data file and extracting header
    val testing         = spark_context.textFile(testing_file).filter(x => ! x.contains("userId"))
    val testing_data    = testing.map( x => {
                            val y = x.split(",")
                            ((y(0).toInt, y(1).toInt), None)
                          })

    // Generating training data by subtracting testing data from it
    val training_data   = input_data.subtractByKey(testing_data)

    //println(training_data)
    // Configuration parameters for training model
    val user_rating     = training_data.map(x => Rating(x._1._1, x._1._2, x._2))
    val rank            = 12
    val number_iter     = 12
    val lambda_val      = 0.1
    val blocks          = 1
    val seed            = 4

    // ALS Model training
    val ALS_model       = ALS.train(ratings = user_rating,
                          rank= rank,
                          iterations= number_iter,
                          lambda= lambda_val,
                          blocks= blocks,
                          seed= seed)

    // Generating testing data
    val user_movies     = testing_data.map(x=> (x._1._1, x._1._2))

    // Predicted values for testing data
    val predicted_vals  = ALS_model.predict(user_movies).map{case Rating(user, product, rating) => ((user, product), rating)}

    // Generating output string as per expected format and writing in Output file
    val output_list     = predicted_vals.sortByKey().map(x=> (x._1._1, x._1._2, x._2)).collect()

    // Generating expected output of the form "User, Movie, Predicted Rating"
    val temp_result     = output_list.mkString("\n")
    val final_result    = temp_result.replace("(", "").replace(")", "").replace(",", ", ")

    // Writing results into output file
    output_file.write(final_result)
    output_file.close()

    // Joining actual ratings and predicted ratings for RMSE calculations
    val real_predicted  = input_data.map(x=> ((x._1._1, x._1._2), x._2)).join(predicted_vals)

    // Calculating and printing counts of error buckets
    val differences     = real_predicted.map(x=> count_item_differences(x._2._1, x._2._2.toFloat)).countByValue()
    for ((key, value) <- differences.toArray.sorted){
      println(key + ": " + value)
    }

    // Calculating MSE and RMSE values and printing RMSE
    val MSE             = real_predicted.map{ x =>
      val err = x._2._1 - x._2._2.toFloat
      err * err
    }.mean()

    val RMSE            = math.sqrt(MSE)
    println("RMSE: " + RMSE)

    // Printing time taken by program
    println("Time: "+ ((System.nanoTime() - start_time) / 1e9d).toInt + " sec")
  }
}

File: Prasad_Bhagwat_ModelBasedCF.py
# Imports required for the program
from pyspark import SparkConf, SparkContext
import time
import sys
import math
from pyspark.mllib.recommendation import ALS, Rating


# Counting error buckets
def count_item_differences(actual, predicted):
    error = abs(actual - predicted)
    if error >= 0 and error < 1:
        return '>=0 and <1'
    elif error >= 1 and error < 2:
        return '>=1 and <2'
    elif error >= 2 and error < 3:
        return '>=2 and <3'
    elif error >= 3 and error < 4:
        return '>=3 and <4'
    else:
        return '>=4'


# Main Function
def main():
    start_time = time.time()

    # Command Line Arguments
    input_file      = sys.argv[1]
    testing_file    = sys.argv[2]

    # Output filename generation
    output_file_name    = "Prasad_Bhagwat_ModelBasedCF.txt"
    output_file         = open(output_file_name, "w")

    # Creating Spark Context
    spark_config    = SparkConf()
    spark_context   = SparkContext(conf=spark_config)
    spark_context.setLogLevel("WARN")

    # Reading input training data file and extracting header
    input           = spark_context.textFile(input_file).filter(lambda x: "userId" not in x)
    input_data      = input.map(lambda x: ((int(x.split(",")[0]), int(x.split(",")[1])), float(x.split(",")[2])))

    # Reading test data file and extracting header
    testing         = spark_context.textFile(testing_file).filter(lambda x: "userId" not in x)
    testing_data    = testing.map(lambda x: ((int(x.split(",")[0]), int(x.split(",")[1])), None))

    # Generating training data by subtracting testing data from it
    training_data   = input_data.subtractByKey(testing_data)

    # Configuration parameters for training model
    user_rating     = training_data.map(lambda x: Rating(int(x[0][0]), int(x[0][1]), float(x[1])))
    rank            = 12    # Can be changed for better performance
    num_iterations  = 12    # Can be changed for better performance
    lambda_value    = 0.1   # Can be changed for better performance
    num_blocks      = 1     # Can be changed for better performance
    seed            = 4     # Can be changed for better performance

    # ALS Model training
    ALS_model       = ALS.train(ratings= user_rating, \
                                rank= rank, \
                                iterations= num_iterations, \
                                lambda_= lambda_value, \
                                blocks= num_blocks, \
                                seed= seed)

    # Generating testing data
    user_movies     = testing_data.map(lambda x: (int(x[0][0]), int(x[0][1])))

    # Predicted values for testing data
    predicted_vals  = ALS_model.predictAll(user_movies).map(lambda x: ((x[0], x[1]), x[2]))

    # Generating output string as per expected format and writing in Output file
    output_list     = predicted_vals.sortByKey().map(lambda x: (x[0][0], x[0][1], x[1]))
    output_result   = "\n".join(map(str, output_list.collect())).replace("(", "").replace(")", "")
    output_file.write(output_result)
    output_file.close()

    # Joining actual ratings and predicted ratings for calculating RMSE
    real_predicted  = input_data.map(lambda x: ((x[0][0], x[0][1]), x[1])).join(predicted_vals)

    # Calculating and printing counts of error buckets
    differences     = real_predicted.map(lambda x : count_item_differences(x[1][0], x[1][1])).countByValue()
    for key in sorted(differences.iterkeys()):
        print "%s: %s" % (key, differences[key])

    # Calculating MSE and RMSE values and printing RMSE
    MSE             = float(real_predicted.map(lambda x: (x[1][0] - x[1][1]) ** 2).mean())
    RMSE            = float(math.sqrt(MSE))
    print "RMSE:", RMSE

    # Printing time taken by the program
    print "Time:", int(time.time() - start_time), "sec"


# Entry point of the program
if __name__ == '__main__':
    main()
